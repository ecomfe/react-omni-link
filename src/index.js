import URITemplates from 'uri-templates';
import omit from 'lodash.omit';
import PropTypes from 'prop-types';
import {Link as RouterLink, NavLink as RouterNavLink} from 'react-router-dom';

export const Link = ({refresh, external, to, activeClassName, ...props}) => {
    if (external) {
        return <a href={to} target="_blank" rel="noopener noreferrer" {...props} />;
    }

    if (refresh) {
        return <a href={to} {...props} />;
    }

    if (activeClassName) {
        return <RouterNavLink to={to} {...props} activeClassName={activeClassName} />;
    }

    return <RouterLink to={to} {...props} />;
};

Link.propTypes = {
    to: PropTypes.string.isRequired,
    refresh: PropTypes.bool,
    external: PropTypes.bool,
    activeClassName: PropTypes.string,
};

Link.defaultProps = {
    refresh: false,
    external: false,
    activeClassName: '',
};

export const ExternalLink = props => <Link external {...props} />;

export const HyperLink = props => <Link refresh {...props} />;

export const createLink = (urlTemplate, defaults = {}) => {
    const template = typeof urlTemplate === 'string' ? new URITemplates(urlTemplate) : urlTemplate;

    if (!template.varNames.length) {
        const to = template.toString();
        const FixedLink = props => <Link to={to} {...defaults} {...props} />;
        return FixedLink;
    }


    const TemplateLink = props => {
        const to = template.fill(props);
        const passDownProps = omit(props, template.varNames);
        return <Link to={to} {...defaults} {...passDownProps} />;
    };
    return TemplateLink;
};
