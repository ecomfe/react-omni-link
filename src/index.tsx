import React, {FC} from 'react';
import uriTemplates, {URITemplate} from 'uri-templates';
import omit from 'lodash.omit';
import PropTypes from 'prop-types';
import {Link as RouterLink, NavLink as RouterNavLink, NavLinkProps as RouterNavLinkProps} from 'react-router-dom';

export const Link: FC<LinkProps> = ({refresh, external, to, activeClassName, ...props}) => {
    if (external) {
        return <a href={to} target="_blank" rel="noopener noreferrer" {...props} />;
    }

    if (refresh) {
        return <a href={to} {...props} />;
    }

    if (activeClassName) {
        return <RouterNavLink to={to} {...props} activeClassName={activeClassName} />;
    }

    return <RouterLink to={to} {...props} />;
};

interface LinkProps<S = {}> extends RouterNavLinkProps<S>{
    to: string;
    refresh?: boolean;
    external?: boolean;
    activeClassName?: string;
}

Link.propTypes = {
    to: PropTypes.string.isRequired,
    refresh: PropTypes.bool,
    external: PropTypes.bool,
    activeClassName: PropTypes.string,
};

Link.defaultProps = {
    refresh: false,
    external: false,
    activeClassName: '',
};

export const ExternalLink: FC<Omit<LinkProps, 'external'>> = props => <Link external {...props} />;

export const HyperLink: FC<Omit<LinkProps, 'refresh'>> = props => <Link refresh {...props} />;

export const createLink = (urlTemplate: URITemplate | string, defaults: Omit<LinkProps, 'to'> = {}) => {
    const template = typeof urlTemplate === 'string' ? uriTemplates(urlTemplate) : urlTemplate;

    if (!template.varNames.length) {
        const to = template.toString();
        const FixedLink: FC<Omit<LinkProps, 'to'>> = props => <Link to={to} {...defaults} {...props} />;
        return FixedLink;
    }


    const TemplateLink: FC<any> = props => {
        const to = template.fill(props);
        const passDownProps = omit(props, template.varNames);
        return <Link to={to} {...defaults} {...passDownProps} />;
    };
    return TemplateLink;
};
