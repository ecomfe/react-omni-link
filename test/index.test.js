import {MemoryRouter as Router} from 'react-router';
import URITemplate from 'uri-templates';
import {createLink, ExternalLink, HyperLink, Link} from '../src';
import {create} from 'react-test-renderer';

test('should create a link by URITemplate', () => {
    const template = new URITemplate('/users/{username}');
    const TemplateLink = createLink(template);

    const component = create(
        <Router>
            <TemplateLink username="admin" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a link by string', () => {
    const UserListLink = createLink('/users');
    const component = create(
        <Router>
            <UserListLink />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a link by string with params', () => {
    const UserProfileLink = createLink('/users/{username}');
    const component = create(
        <Router>
            <UserProfileLink username="admin" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a link by string with encoded params', () => {
    const UserProfileLink = createLink('/users/{username}');
    const component = create(
        <Router>
            <UserProfileLink username="admin/" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a external link by string', () => {
    const AdminConsoleLink = createLink('/console', {external: true});

    const component = create(
        <Router>
            <AdminConsoleLink />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a link by string without encoding', () => {
    const DocumentViewLink = createLink('/documents/{+path}');
    const component = create(
        <Router>
            <DocumentViewLink path="path/to/file?query=1" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a link by string with query', () => {
    const DocumentEditLink = createLink('/documents/{+path}/edit{?query*}');
    const query = {a: '1', b: ['2', '3']};
    const component = create(
        <Router>
            <DocumentEditLink path="path/to/file" query={query} />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a external link by shortcut', () => {
    const component = create(
        <Router>
            <ExternalLink to="/users" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a refresh link by shortcut', () => {
    const component = create(
        <Router>
            <HyperLink to="/users" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});

test('should create a actived nav link', () => {
    const component = create(
        <Router initialEntries={['/users']}>
            <Link to="/users" activeClassName="active" />
        </Router>
    );

    expect(component.toJSON()).toMatchSnapshot();
});
