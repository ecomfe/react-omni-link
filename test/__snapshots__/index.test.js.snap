// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a actived nav link 1`] = `
<a
  aria-current="page"
  className="active"
  href="/users"
  onClick={[Function]}
  style={Object {}}
/>
`;

exports[`should create a external link by shortcut 1`] = `
<a
  href="/users"
  rel="noopener noreferrer"
  target="_blank"
/>
`;

exports[`should create a external link by string 1`] = `
<a
  href="/console"
  rel="noopener noreferrer"
  target="_blank"
/>
`;

exports[`should create a link by URITemplate 1`] = `
<a
  href="/users/admin"
  onClick={[Function]}
/>
`;

exports[`should create a link by string 1`] = `
<a
  href="/users"
  onClick={[Function]}
/>
`;

exports[`should create a link by string with encoded params 1`] = `
<a
  href="/users/admin%2F"
  onClick={[Function]}
/>
`;

exports[`should create a link by string with params 1`] = `
<a
  href="/users/admin"
  onClick={[Function]}
/>
`;

exports[`should create a link by string with query 1`] = `
<a
  href="/documents/path/to/file/edit?a=1&b=2%2C3"
  onClick={[Function]}
/>
`;

exports[`should create a link by string without encoding 1`] = `
<a
  href="/documents/path/to/file?query=1"
  onClick={[Function]}
/>
`;

exports[`should create a refresh link by shortcut 1`] = `
<a
  href="/users"
/>
`;
